server/service/service_osquery.go

```diff
@@ -447,15 +448,92 @@ func (svc *Service) detailQueriesForHost(ctx context.Context, host *fleet.Host)
-		jitter = time.Duration((int64(hostID) + svc.getJitterSeed()) % maxJitter)
+	svc.jitterMu.Lock()
+	defer svc.jitterMu.Unlock()
+
+	if svc.jitterH[interval] == nil {
+		svc.jitterH[interval] = &jitterHashTable{
+			maxCapacity: 1,
+			bucketCount: int(int64(svc.config.Osquery.MaxJitterPercent) * int64(interval.Minutes()) / 100.0),
+			buckets:     make(map[int]int),
+			cache:       make(map[uint]time.Duration),
+		}
+		if svc.jitterH[interval].bucketCount == 0 {
+			svc.jitterH[interval].bucketCount = 1
+		}
+		level.Debug(svc.logger).Log("jitter", "created", "bucketCount", svc.jitterH[interval].bucketCount)
 	}
+
+	jitter := svc.jitterH[interval].jitterForHost(hostID)
 	cutoff := svc.clock.Now().Add(-(interval + jitter))
 	return lastUpdated.Before(cutoff)
 }
 
+// jitterHashTable implements a data structure that allows a fleet to generate a static jitter value
+// that is properly balanced. Balance in this context means that hosts would be distributed uniformly
+// across the total jitter time so there are no spikes.
+// The way this structure works is as follows:
+// Given an amount of buckets, we want to place hosts in buckets evenly. So we don't want bucket 0 to
+// have 1000 hosts, and all the other buckets 0. If there were 1000 buckets, and 1000 hosts, we should
+// end up with 1 per bucket.
+// The total amount of online hosts is unknown, so first it assumes that amount of buckets >= amount
+// of total hosts (maxCapacity of 1 per bucket). Once we have more hosts than buckets, then we
+// increase the maxCapacity by 1 for all buckets, and start placing hosts.
+// Hosts that have been placed in a bucket remain in that bucket for as long as the fleet instance is
+// running.
+// The preferred bucket for a host is the one at (host id % bucketCount). If that bucket is full, the
+// next one will be tried. If all buckets are full, then capacity gets increased and the bucket
+// selection process restarts.
+// Once a bucket is found, the index for the bucket (going from 0 to bucketCount) will be the amount of
+// minutes added to the host check in time.
+// For example: at a 1hr interval, and the default 10% max jitter percent. That allows hosts to
+// distribute within 6 minutes around the hour mark. We would have 6 buckets in that case.
+// In the worse possible case that all hosts start at the same time, max jitter percent can be set to
+// 100, and this method will distribute hosts evenly.
+// The main caveat of this approach is that it works at the fleet instance. So depending on what
+// instance gets chosen by the load balancer, the jitter might be different. However, load tests have
+// shown that
+type jitterHashTable struct {
+	mu          sync.Mutex
+	maxCapacity int
+	bucketCount int
+	buckets     map[int]int
+	cache       map[uint]time.Duration
+}
+
+func (jh *jitterHashTable) jitterForHost(hostID uint) time.Duration {
+	// if no jitter is configured just return 0
+	if jh.bucketCount == 0 {
```

Comment:
Should this be `== 1`?

When you create the `jitterHashTable` you do:
```go
		if svc.jitterH[interval].bucketCount == 0 {
			svc.jitterH[interval].bucketCount = 1
		}
```

Reply:
Good point. Basically, when you mentioned testing without jitter, I wrote the test directly with the jitter and things exploded. So I added that early check. If bucketCount is 1, then everything will fit in index 0 quickly enough... I suppose I could check for `<=1` instead here, as it'll always be zero then. Does that sound reasonable to you?

Reply:
+1, also: worth defining a `newJitterHashTable(interval, maxJitterPercent)` that takes care of setting the struct values properly.